File: backend/app/services/settings_service.py
from __future__ import annotations

import base64
import hashlib
from dataclasses import dataclass
import os

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.models.setting import SecretSetting, Setting


DEFAULT_SETTINGS = {
    "daily_digest_time": "12:00",
    "weekly_digest_day_of_week": "sat",
    "weekly_digest_time": "23:30",
    "recipient_email": None,
    "sender_identity": None,
    "smtp_host": None,
    "smtp_port": "587",
    "smtp_username": None,
    "smtp_security_mode": "starttls",
    "daily_digest_enabled": "true",
    "daily_digest_catch_up_enabled": "true",
    "weekly_digest_enabled": "false",
    "weekly_digest_catch_up_enabled": "true",
}


@dataclass(frozen=True)
class AppSettings:
    daily_digest_time: str
    weekly_digest_day_of_week: str
    weekly_digest_time: str
    recipient_email: str | None
    sender_identity: str | None
    smtp_host: str | None
    smtp_port: int
    smtp_username: str | None
    smtp_security_mode: str
    daily_digest_enabled: bool
    daily_digest_catch_up_enabled: bool
    weekly_digest_enabled: bool
    weekly_digest_catch_up_enabled: bool
    smtp_password_configured: bool


@dataclass(frozen=True)
class AppSettingsUpdate:
    daily_digest_time: str
    weekly_digest_day_of_week: str
    weekly_digest_time: str
    recipient_email: str | None
    sender_identity: str | None
    smtp_host: str | None
    smtp_port: int
    smtp_username: str | None
    smtp_security_mode: str
    daily_digest_enabled: bool
    daily_digest_catch_up_enabled: bool
    weekly_digest_enabled: bool
    weekly_digest_catch_up_enabled: bool
    smtp_password: str | None


def _as_bool(value: str | None, default: bool) -> bool:
    if value is None:
        return default
    return value.lower() == "true"


def _derive_keystream(*, master_key: str, nonce: bytes, length: int) -> bytes:
    seed = hashlib.sha256(master_key.encode("utf-8") + nonce).digest()
    blocks: list[bytes] = []
    counter = 0
    while sum(len(block) for block in blocks) < length:
        blocks.append(hashlib.sha256(seed + counter.to_bytes(4, "big")).digest())
        counter += 1
    return b"".join(blocks)[:length]


def encrypt_secret(plaintext: str, master_key: str) -> str:
    plaintext_bytes = plaintext.encode("utf-8")
    nonce = os.urandom(16)
    keystream = _derive_keystream(master_key=master_key, nonce=nonce, length=len(plaintext_bytes))
    ciphertext = bytes(left ^ right for left, right in zip(plaintext_bytes, keystream, strict=True))
    return base64.urlsafe_b64encode(nonce + ciphertext).decode("utf-8")


def decrypt_secret(ciphertext: str, master_key: str) -> str:
    payload = base64.urlsafe_b64decode(ciphertext.encode("utf-8"))
    nonce = payload[:16]
    encrypted_bytes = payload[16:]
    keystream = _derive_keystream(master_key=master_key, nonce=nonce, length=len(encrypted_bytes))
    plaintext = bytes(left ^ right for left, right in zip(encrypted_bytes, keystream, strict=True))
    return plaintext.decode("utf-8")


def load_settings(session: Session) -> AppSettings:
    stored = {row.key: row.value for row in session.execute(select(Setting)).scalars()}
    secret_keys = {row.key for row in session.execute(select(SecretSetting)).scalars()}
    return AppSettings(
        daily_digest_time=stored.get("daily_digest_time") or DEFAULT_SETTINGS["daily_digest_time"],
        weekly_digest_day_of_week=stored.get("weekly_digest_day_of_week")
        or DEFAULT_SETTINGS["weekly_digest_day_of_week"],
        weekly_digest_time=stored.get("weekly_digest_time") or DEFAULT_SETTINGS["weekly_digest_time"],
        recipient_email=stored.get("recipient_email"),
        sender_identity=stored.get("sender_identity"),
        smtp_host=stored.get("smtp_host"),
        smtp_port=int(stored.get("smtp_port") or DEFAULT_SETTINGS["smtp_port"]),
        smtp_username=stored.get("smtp_username"),
        smtp_security_mode=stored.get("smtp_security_mode") or DEFAULT_SETTINGS["smtp_security_mode"],
        daily_digest_enabled=_as_bool(stored.get("daily_digest_enabled"), True),
        daily_digest_catch_up_enabled=_as_bool(stored.get("daily_digest_catch_up_enabled"), True),
        weekly_digest_enabled=_as_bool(stored.get("weekly_digest_enabled"), False),
        weekly_digest_catch_up_enabled=_as_bool(stored.get("weekly_digest_catch_up_enabled"), True),
        smtp_password_configured="smtp_password" in secret_keys,
    )


def _upsert_setting(session: Session, key: str, value: str | None) -> None:
    existing = session.scalar(select(Setting).where(Setting.key == key))
    if existing is None:
        session.add(Setting(key=key, value=value))
        return
    existing.value = value


def _upsert_secret_setting(session: Session, key: str, encrypted_value: str) -> None:
    existing = session.scalar(select(SecretSetting).where(SecretSetting.key == key))
    if existing is None:
        session.add(SecretSetting(key=key, encrypted_value=encrypted_value))
        return
    existing.encrypted_value = encrypted_value


def save_settings(session: Session, payload: AppSettingsUpdate, master_key: str) -> AppSettings:
    _upsert_setting(session, "daily_digest_time", payload.daily_digest_time)
    _upsert_setting(session, "weekly_digest_day_of_week", payload.weekly_digest_day_of_week)
    _upsert_setting(session, "weekly_digest_time", payload.weekly_digest_time)
    _upsert_setting(session, "recipient_email", payload.recipient_email)
    _upsert_setting(session, "sender_identity", payload.sender_identity)
    _upsert_setting(session, "smtp_host", payload.smtp_host)
    _upsert_setting(session, "smtp_port", str(payload.smtp_port))
    _upsert_setting(session, "smtp_username", payload.smtp_username)
    _upsert_setting(session, "smtp_security_mode", payload.smtp_security_mode)
    _upsert_setting(session, "daily_digest_enabled", str(payload.daily_digest_enabled).lower())
    _upsert_setting(session, "daily_digest_catch_up_enabled", str(payload.daily_digest_catch_up_enabled).lower())
    _upsert_setting(session, "weekly_digest_enabled", str(payload.weekly_digest_enabled).lower())
    _upsert_setting(session, "weekly_digest_catch_up_enabled", str(payload.weekly_digest_catch_up_enabled).lower())
    if payload.smtp_password:
        _upsert_secret_setting(session, "smtp_password", encrypt_secret(payload.smtp_password, master_key))
    session.flush()
    return load_settings(session)


def get_smtp_password(session: Session, master_key: str) -> str | None:
    secret_setting = session.scalar(select(SecretSetting).where(SecretSetting.key == "smtp_password"))
    if secret_setting is None:
        return None
    return decrypt_secret(secret_setting.encrypted_value, master_key)


def set_last_daily_digest_sent_at(session: Session, scheduled_for_iso: str) -> None:
    _upsert_setting(session, "last_daily_digest_sent_at", scheduled_for_iso)


def get_last_daily_digest_sent_at(session: Session) -> str | None:
    row = session.scalar(select(Setting).where(Setting.key == "last_daily_digest_sent_at"))
    return row.value if row is not None else None


def set_last_weekly_digest_sent_at(session: Session, scheduled_for_iso: str) -> None:
    _upsert_setting(session, "last_weekly_digest_sent_at", scheduled_for_iso)


def get_last_weekly_digest_sent_at(session: Session) -> str | None:
    row = session.scalar(select(Setting).where(Setting.key == "last_weekly_digest_sent_at"))
    return row.value if row is not None else None


def set_last_observability_report_sent_at(session: Session, scheduled_for_iso: str) -> None:
    _upsert_setting(session, "last_observability_report_sent_at", scheduled_for_iso)


def get_last_observability_report_sent_at(session: Session) -> str | None:
    row = session.scalar(select(Setting).where(Setting.key == "last_observability_report_sent_at"))
    return row.value if row is not None else None

File: backend-java/src/main/java/com/goodnews/backendjava/service/SettingsService.java
package com.goodnews.backendjava.service;

import com.goodnews.backendjava.api.dto.SettingsDtos;
import com.goodnews.backendjava.config.GoodNewsProperties;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.util.Base64;
import java.util.LinkedHashMap;
import java.util.Map;
import org.springframework.r2dbc.core.DatabaseClient;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@Service
public class SettingsService {

    static final String DEFAULT_SUMMARY_INSTRUCTIONS = """
        a detailed summary in Russian (Кириллица), 4-6 sentences (roughly 60-120 words). Cover what the article is about, its key points, arguments or findings, and the concrete takeaway — enough that the reader understands the substance without opening the article. Do not just rephrase the title; add the specifics from the body. Use only Russian words — no Latin, no code, no transliteration. If you cannot write proper Russian, use "".
        """.strip();
    static final String DEFAULT_VERDICT_REASON_INSTRUCTIONS = """
        1 sentence in ENGLISH explaining why THIS reader (given the reader preference profile above, when present) would or would not want to read this. MUST be in English. No Russian.
        """.strip();

    private static final Map<String, String> DEFAULT_SETTINGS = Map.ofEntries(
        Map.entry("daily_digest_time", "12:00"),
        Map.entry("weekly_digest_day_of_week", "sat"),
        Map.entry("weekly_digest_time", "23:30"),
        Map.entry("smtp_port", "587"),
        Map.entry("smtp_security_mode", "starttls"),
        Map.entry("daily_digest_enabled", "true"),
        Map.entry("daily_digest_catch_up_enabled", "true"),
        Map.entry("weekly_digest_enabled", "false"),
        Map.entry("weekly_digest_catch_up_enabled", "true"),
        Map.entry("analysis_summary_prompt", DEFAULT_SUMMARY_INSTRUCTIONS),
        Map.entry("analysis_verdict_reason_prompt", DEFAULT_VERDICT_REASON_INSTRUCTIONS)
    );

    private final DatabaseClient databaseClient;
    private final GoodNewsProperties properties;

    public SettingsService(DatabaseClient databaseClient, GoodNewsProperties properties) {
        this.databaseClient = databaseClient;
        this.properties = properties;
    }

    public Mono<SettingsDtos.SettingsResponse> getSettingsResponse() {
        return loadSettings().map(this::toResponse);
    }

    @Transactional
    public Mono<SettingsDtos.SettingsResponse> updateSettings(SettingsDtos.SettingsUpdateRequest request) {
        AppSettingsUpdate update = new AppSettingsUpdate(
            request.normalizedDailyDigestTime(),
            request.normalizedWeeklyDigestDayOfWeek(),
            request.normalizedWeeklyDigestTime(),
            request.recipient_email(),
            request.sender_identity(),
            request.smtp_host(),
            request.smtp_port(),
            request.smtp_username(),
            request.smtp_security_mode(),
            request.daily_digest_enabled(),
            request.daily_digest_catch_up_enabled(),
            request.weekly_digest_enabled(),
            request.weekly_digest_catch_up_enabled(),
            request.smtp_password(),
            request.analysis_summary_prompt(),
            request.analysis_verdict_reason_prompt()
        );
        return saveSettings(update).map(this::toResponse);
    }

    public Mono<AppSettings> loadSettings() {
        Mono<Map<String, String>> storedSettings = databaseClient.sql("SELECT key, value FROM settings")
            .map((row, metadata) -> new StoredSettingRow(
                row.get("key", String.class),
                row.get("value", String.class)
            ))
            .all()
            .collectMap(StoredSettingRow::key, StoredSettingRow::value, LinkedHashMap::new);

        Mono<Map<String, Boolean>> secretSettings = databaseClient.sql("SELECT key FROM secret_settings")
            .map((row, metadata) -> row.get("key", String.class))
            .all()
            .collectMap(key -> key, key -> true, LinkedHashMap::new);

        return Mono.zip(storedSettings, secretSettings)
            .map(tuple -> {
                Map<String, String> stored = tuple.getT1();
                Map<String, Boolean> secrets = tuple.getT2();
                return new AppSettings(
                    stored.getOrDefault("daily_digest_time", DEFAULT_SETTINGS.get("daily_digest_time")),
                    stored.getOrDefault("weekly_digest_day_of_week", DEFAULT_SETTINGS.get("weekly_digest_day_of_week")),
                    stored.getOrDefault("weekly_digest_time", DEFAULT_SETTINGS.get("weekly_digest_time")),
                    stored.get("recipient_email"),
                    stored.get("sender_identity"),
                    stored.get("smtp_host"),
                    Integer.parseInt(stored.getOrDefault("smtp_port", DEFAULT_SETTINGS.get("smtp_port"))),
                    stored.get("smtp_username"),
                    stored.getOrDefault("smtp_security_mode", DEFAULT_SETTINGS.get("smtp_security_mode")),
                    asBoolean(stored.get("daily_digest_enabled"), true),
                    asBoolean(stored.get("daily_digest_catch_up_enabled"), true),
                    asBoolean(stored.get("weekly_digest_enabled"), false),
                    asBoolean(stored.get("weekly_digest_catch_up_enabled"), true),
                    secrets.containsKey("smtp_password"),
                    defaultIfBlank(stored.get("analysis_summary_prompt"), DEFAULT_SETTINGS.get("analysis_summary_prompt")),
                    defaultIfBlank(
                        stored.get("analysis_verdict_reason_prompt"),
                        DEFAULT_SETTINGS.get("analysis_verdict_reason_prompt")
                    )
                );
            });
    }

    public Mono<AppSettings> saveSettings(AppSettingsUpdate update) {
        Mono<Void> writes = Flux.concat(
                upsertSetting("daily_digest_time", update.dailyDigestTime()),
                upsertSetting("weekly_digest_day_of_week", update.weeklyDigestDayOfWeek()),
                upsertSetting("weekly_digest_time", update.weeklyDigestTime()),
                upsertSetting("recipient_email", update.recipientEmail()),
                upsertSetting("sender_identity", update.senderIdentity()),
                upsertSetting("smtp_host", update.smtpHost()),
                upsertSetting("smtp_port", Integer.toString(update.smtpPort())),
                upsertSetting("smtp_username", update.smtpUsername()),
                upsertSetting("smtp_security_mode", update.smtpSecurityMode()),
                upsertSetting("daily_digest_enabled", Boolean.toString(update.dailyDigestEnabled()).toLowerCase()),
                upsertSetting(
                    "daily_digest_catch_up_enabled",
                    Boolean.toString(update.dailyDigestCatchUpEnabled()).toLowerCase()
                ),
                upsertSetting("weekly_digest_enabled", Boolean.toString(update.weeklyDigestEnabled()).toLowerCase()),
                upsertSetting(
                    "weekly_digest_catch_up_enabled",
                    Boolean.toString(update.weeklyDigestCatchUpEnabled()).toLowerCase()
                ),
                upsertSetting("analysis_summary_prompt", blankToNull(update.analysisSummaryPrompt())),
                upsertSetting("analysis_verdict_reason_prompt", blankToNull(update.analysisVerdictReasonPrompt()))
            )
            .then();

        Mono<Void> secretWrite = hasText(update.smtpPassword())
            ? upsertSecretSetting("smtp_password", encryptSecret(update.smtpPassword(), requireAppMasterKey()))
            : Mono.empty();

        return writes.then(secretWrite).then(loadSettings());
    }

    public Mono<String> getSmtpPassword() {
        return databaseClient.sql("SELECT encrypted_value FROM secret_settings WHERE key = :key")
            .bind("key", "smtp_password")
            .map((row, metadata) -> row.get("encrypted_value", String.class))
            .one()
            .map(encrypted -> decryptSecret(encrypted, requireAppMasterKey()));
    }

    static String encryptSecret(String plaintext, String masterKey) {
        byte[] nonce = new byte[16];
        new SecureRandom().nextBytes(nonce);
        return encryptSecretWithNonce(plaintext, masterKey, nonce);
    }

    static String encryptSecretWithNonce(String plaintext, String masterKey, byte[] nonce) {
        byte[] plaintextBytes = plaintext.getBytes(StandardCharsets.UTF_8);
        byte[] keystream = deriveKeystream(masterKey, nonce, plaintextBytes.length);
        byte[] ciphertext = xor(plaintextBytes, keystream);
        byte[] payload = ByteBuffer.allocate(nonce.length + ciphertext.length)
            .put(nonce)
            .put(ciphertext)
            .array();
        return Base64.getUrlEncoder().encodeToString(payload);
    }

    static String decryptSecret(String ciphertext, String masterKey) {
        byte[] payload = Base64.getUrlDecoder().decode(ciphertext.getBytes(StandardCharsets.UTF_8));
        byte[] nonce = java.util.Arrays.copyOfRange(payload, 0, 16);
        byte[] encryptedBytes = java.util.Arrays.copyOfRange(payload, 16, payload.length);
        byte[] keystream = deriveKeystream(masterKey, nonce, encryptedBytes.length);
        return new String(xor(encryptedBytes, keystream), StandardCharsets.UTF_8);
    }

    private static byte[] deriveKeystream(String masterKey, byte[] nonce, int length) {
        byte[] seed = sha256(concat(masterKey.getBytes(StandardCharsets.UTF_8), nonce));
        ByteBuffer keystream = ByteBuffer.allocate(length);
        int counter = 0;
        while (keystream.position() < length) {
            byte[] block = sha256(concat(seed, ByteBuffer.allocate(4).putInt(counter).array()));
            keystream.put(block, 0, Math.min(block.length, keystream.remaining()));
            counter++;
        }
        return keystream.array();
    }

    private Mono<Void> upsertSetting(String key, String value) {
        DatabaseClient.GenericExecuteSpec spec = databaseClient.sql("""
                INSERT INTO settings (key, value, created_at, updated_at)
                VALUES (:key, :value, CURRENT_TIMESTAMP, CURRENT_TIMESTAMP)
                ON CONFLICT (key) DO UPDATE
                SET value = EXCLUDED.value,
                    updated_at = CURRENT_TIMESTAMP
                """)
            .bind("key", key);
        spec = bindNullable(spec, "value", value, String.class);
        return spec.fetch().rowsUpdated().then();
    }

    private Mono<Void> upsertSecretSetting(String key, String encryptedValue) {
        return databaseClient.sql("""
                INSERT INTO secret_settings (key, encrypted_value, updated_at)
                VALUES (:key, :encryptedValue, CURRENT_TIMESTAMP)
                ON CONFLICT (key) DO UPDATE
                SET encrypted_value = EXCLUDED.encrypted_value,
                    updated_at = CURRENT_TIMESTAMP
                """)
            .bind("key", key)
            .bind("encryptedValue", encryptedValue)
            .fetch()
            .rowsUpdated()
            .then();
    }

    private SettingsDtos.SettingsResponse toResponse(AppSettings settings) {
        return new SettingsDtos.SettingsResponse(
            settings.dailyDigestTime(),
            settings.weeklyDigestDayOfWeek(),
            settings.weeklyDigestTime(),
            properties.observability().dashboardUrl(),
            settings.recipientEmail(),
            settings.senderIdentity(),
            settings.smtpHost(),
            settings.smtpPort(),
            settings.smtpUsername(),
            settings.smtpSecurityMode(),
            settings.dailyDigestEnabled(),
            settings.dailyDigestCatchUpEnabled(),
            settings.weeklyDigestEnabled(),
            settings.weeklyDigestCatchUpEnabled(),
            settings.smtpPasswordConfigured(),
            settings.analysisSummaryPrompt(),
            settings.analysisVerdictReasonPrompt()
        );
    }

    private String requireAppMasterKey() {
        String appMasterKey = properties.email().appMasterKey();
        if (hasText(appMasterKey)) {
            return appMasterKey;
        }
        throw new IllegalStateException(
            "Missing app master key contract GOOD_NEWS_APP_MASTER_KEY; set it explicitly for SMTP secret encryption."
        );
    }

    private static DatabaseClient.GenericExecuteSpec bindNullable(
        DatabaseClient.GenericExecuteSpec spec,
        String name,
        String value,
        Class<String> type
    ) {
        return value == null ? spec.bindNull(name, type) : spec.bind(name, value);
    }

    private static boolean asBoolean(String value, boolean defaultValue) {
        if (value == null) {
            return defaultValue;
        }
        return "true".equalsIgnoreCase(value);
    }

    private static String blankToNull(String value) {
        if (value == null || value.isBlank()) {
            return null;
        }
        return value;
    }

    private static String defaultIfBlank(String value, String defaultValue) {
        if (value == null || value.isBlank()) {
            return defaultValue;
        }
        return value;
    }

    private static boolean hasText(String value) {
        return value != null && !value.isBlank();
    }

    private static byte[] concat(byte[] left, byte[] right) {
        ByteBuffer buffer = ByteBuffer.allocate(left.length + right.length);
        return buffer.put(left).put(right).array();
    }

    private static byte[] xor(byte[] left, byte[] right) {
        byte[] result = new byte[left.length];
        for (int index = 0; index < left.length; index++) {
            result[index] = (byte) (left[index] ^ right[index]);
        }
        return result;
    }

    private static byte[] sha256(byte[] value) {
        try {
            return MessageDigest.getInstance("SHA-256").digest(value);
        } catch (NoSuchAlgorithmException exception) {
            throw new IllegalStateException("Missing SHA-256 support", exception);
        }
    }

    public record AppSettings(
        String dailyDigestTime,
        String weeklyDigestDayOfWeek,
        String weeklyDigestTime,
        String recipientEmail,
        String senderIdentity,
        String smtpHost,
        int smtpPort,
        String smtpUsername,
        String smtpSecurityMode,
        boolean dailyDigestEnabled,
        boolean dailyDigestCatchUpEnabled,
        boolean weeklyDigestEnabled,
        boolean weeklyDigestCatchUpEnabled,
        boolean smtpPasswordConfigured,
        String analysisSummaryPrompt,
        String analysisVerdictReasonPrompt
    ) {}

    public record AppSettingsUpdate(
        String dailyDigestTime,
        String weeklyDigestDayOfWeek,
        String weeklyDigestTime,
        String recipientEmail,
        String senderIdentity,
        String smtpHost,
        int smtpPort,
        String smtpUsername,
        String smtpSecurityMode,
        boolean dailyDigestEnabled,
        boolean dailyDigestCatchUpEnabled,
        boolean weeklyDigestEnabled,
        boolean weeklyDigestCatchUpEnabled,
        String smtpPassword,
        String analysisSummaryPrompt,
        String analysisVerdictReasonPrompt
    ) {}

    private record StoredSettingRow(String key, String value) {}
}
